### PR TITLE
perf: run parse_compilation_units once

### PR DESCRIPTION
**Pull request by jchavarri**
`parse_compilation_units` can be expensive when there are a lot of modules in the map. This PR makes sure it only runs once per lib (or other stanzas).

Some measurements for a few actions in Ahrefs repo. All actions are "noops" (the command was run previously successfully so no user actions should be executed):

| Command | `main`  | This branch |
| ------------- | ------------- |  ------------- |
| `dune exec gen_esgg`  | `real	0m10.121s, user	0m9.677s, sys	0m0.440s` | `real	0m1.372s, user	0m1.157s, sys	0m0.216s`  |
| `dune build @fe.all`  | `real	0m32.039s, user	0m30.391s, sys	0m1.685s` | `real	0m12.716s, user	0m11.615s, sys	0m1.119s`  |
| `dune build @gen_dune`  | `real	0m25.436s, user	0m24.253s, sys	0m1.223s` | `real	0m5.407s, user	0m4.908s, sys	0m0.504s`  |

### TODO

- [x] Create public benchmark to test this PR (#7189, #7198)

### CODE REVIEW

**Comment by rgrinberg** [comment]
Fix looks right. I'm wondering if we could simplify things even further though. How about saving the object map as part of `Modules.t`? This would remove a lot of the staging as well.

**Comment by jchavarri** [comment]
@rgrinberg how do you think the types should be modified in order to move in that direction? Something like this? 

```
type t =
  { obj_map : Module.t Module_name.Unique.Map.t
  ; modules : modules
  }

and modules =
  | Singleton of Module.t
  | Unwrapped of Unwrapped.t
  | Wrapped of Wrapped.t
  | Impl of impl
  | Stdlib of Stdlib.t

and impl =
  { impl : t
  ; vlib : t
  }
```

**Comment by jchavarri** [comment]
Another thing I wonder about the idea of saving the object map as part of `Modules.t` is if there can be cases where this object map would be created unnecessarily, in a wasteful way 🤔

**Comment by rgrinberg** [comment]
> how do you think the types should be modified in order to move in that direction? Something like this?

Yeah that sounds good.

> Another thing I wonder about the idea of saving the object map as part of Modules.t is if there can be cases where this object map would be created unnecessarily, in a wasteful way 🤔

You could make it lazy to avoid that.

File: src/dune_rules/modules.ml [hunk 1]
```diff
@@ -741,39 +746,93 @@ and impl =
...
   }
 
+module Sourced_module = struct
+  type t =
+    | Normal of Module.t
+    | Imported_from_vlib of Module.t
+    | Impl_of_virtual_module of Module.t Ml_kind.Dict.t
+end
+
+let rec obj_map :
```
Comment (jchavarri) [inline src/dune_rules/modules.ml:756]: This function and `obj_map_build` have not changed, just moved up due to scoping issues.

File: src/dune_rules/vimpl.ml [hunk 2]
```diff
@@ -30,13 +30,7 @@ let make ~vlib ~impl ~vlib_modules ~vlib_foreign_objects =
...
     Mode.cm_kind (if byte then Byte else Native)
   in
-  let vlib_obj_map =
-    Modules.obj_map vlib_modules ~f:(function
-      | Normal m -> m
-      | _ -> assert false)
-    |> Module.Obj_map.fold ~init:Module_name.Unique.Map.empty ~f:(fun m acc ->
-           Module_name.Unique.Map.add_exn acc (Module.obj_name m) m)
-  in
+  let vlib_obj_map = Modules.unique_map vlib_modules in
```
Comment (jchavarri) [inline src/dune_rules/vimpl.ml:33]: ⚠️ After this change, the assertion above will never be triggered anymore, as `Modules.unique_map` permits both `Imported_from_vlib` and `Impl_of_virtual_module` values. Should we add a special case for vlibs?

**Review by jchavarri — COMMENTED** [review]
@rgrinberg please take a look to see if this is going in the right direction, thanks!

File: src/dune_rules/vimpl.ml (hunk 2 above)
Comment (rgrinberg) [reply src/dune_rules/vimpl.ml:33]: I'm thinking, do we even need to compute this map right now? We're already saving `vlib_modules`, what about getting rid of this field?

File: src/dune_rules/modules.mli [hunk 3]
```diff
@@ -92,7 +92,7 @@ module Sourced_module : sig
     | Impl_of_virtual_module of Module.t Ml_kind.Dict.t
 end
 
-val obj_map : t -> f:(Sourced_module.t -> 'a) -> 'a Module.Obj_map.t
+val unique_map : t -> Module.t Module_name.Unique.Map.t
```
Comment (rgrinberg) [inline src/dune_rules/modules.mli:95]: Let's stick to `obj_map`

File: src/dune_rules/modules.ml [hunk 4]
```diff
@@ -741,39 +746,93 @@ and impl =
...
     ; ( "wrapped"
       , let+ w = Wrapped.decode ~src_dir in
-        Wrapped w )
+        let modules = Wrapped w in
+        with_unique_map modules )
     ; ( "stdlib"
       , let+ stdlib = Stdlib.decode ~src_dir in
-        Stdlib stdlib )
+        let modules = Stdlib stdlib in
+        with_unique_map modules )
```
Comment (rgrinberg) [inline src/dune_rules/modules.ml:830]: What about just:

```
let+ modules =  sum ... in
with_unique_map modules
```

File: src/dune_rules/ocamldep.mli [hunk 5]
```diff
@@ -32,6 +32,8 @@ val read_deps_of :
   -> Module.t
   -> Module.t list Action_builder.t
 
+val parse_compilation_units : modules:Modules.t -> string list -> Module.t list
```
Comment (rgrinberg) [inline src/dune_rules/ocamldep.mli:35]: is this needed?

File: src/dune_rules/modules.ml [hunk 6]
```diff
@@ -1028,73 +1105,40 @@ let rec fold_user_written t ~f ~init =
...
-    | Stdlib w -> Stdlib (Stdlib.map w ~f)
-    | Wrapped w -> Wrapped (Wrapped.map w ~f)
-    | Impl w -> Impl { w with impl = loop w.impl }
+  let rec loop t =
+    match t.modules with
+    | Singleton m -> with_unique_map (Singleton (f m))
+    | Unwrapped m -> with_unique_map (Unwrapped (Unwrapped.map ~f m))
+    | Stdlib w -> with_unique_map (Stdlib (Stdlib.map w ~f))
+    | Wrapped w -> with_unique_map (Wrapped (Wrapped.map w ~f))
+    | Impl w -> with_unique_map (Impl { w with impl = loop w.impl })
```
Comment (rgrinberg) [inline src/dune_rules/modules.ml:1134]: perhaps `with_unique_map (loop ..)`

File: src/dune_rules/modules.ml [hunk 7]
```diff
@@ -1028,73 +1105,40 @@ let rec fold_user_written t ~f ~init =
...
   | Wrapped
       ({ group; wrapped_compat = _; wrapped = _; toplevel_module = _ } as w) ->
     let+ group = Group.Memo_traversals.parallel_map group ~f in
-    Wrapped { w with group }
+    with_unique_map (Wrapped { w with group })
   | Impl t ->
-    let+ vlib = map_user_written t.vlib ~f in
-    Impl { t with vlib }
+    let+ modules = map_user_written t.vlib ~f in
+    with_unique_map (Impl { t with vlib = modules })
```
Comment (rgrinberg) [inline src/dune_rules/modules.ml:1124]: would be nice to move `with_unique_map` to the end.

File: src/dune_rules/ocamldep.mli (hunk 5 above)
Comment (jchavarri) [reply src/dune_rules/ocamldep.mli]: No. Removed in https://github.com/ocaml/dune/pull/7187/commits/955fab5f9bced5987da8265b16f4ca2bcafcfd03.

File: src/dune_rules/vimpl.ml (hunk 2 above)
Comment (jchavarri) [reply src/dune_rules/vimpl.ml]: We don't need to, updated record in https://github.com/ocaml/dune/pull/7187/commits/955fab5f9bced5987da8265b16f4ca2bcafcfd03. My previous comment about the assertion going away still applies, I understand it's not a big deal.

**Comment by jchavarri** [comment]
@rgrinberg thanks for the review, I think I tackled all your comments, the code looks better now.

If it looks good to merge, before doing so please take a look at #7198. I would like to merge that benchmark PR first, so that we can make measurements compare and understand better the impact in warm builds, thanks!

File: src/dune_rules/modules.mli [hunk 8]
```diff
@@ -92,7 +92,7 @@ module Sourced_module : sig
     | Impl_of_virtual_module of Module.t Ml_kind.Dict.t
 end
 
-val obj_map : t -> f:(Sourced_module.t -> 'a) -> 'a Module.Obj_map.t
+val obj_map : t -> Module.t Module_name.Unique.Map.t
```
Comment (rgrinberg) [inline src/dune_rules/modules.mli:95]: Should we just make this valued by `Sourced_module.t`? It doesn't really hurt for `ocamldep`, and it will remove further redundant computations.

Comment (jchavarri) [reply src/dune_rules/modules.mli]: Do you mean the type being 

```
val obj_map : t -> Sourced_module.t Module_name.Unique.Map.t
```
?
I tried something like that in https://github.com/ocaml/dune/pull/7187/commits/c8e81be6742744b558079a4fafc5a0d1ba6b3e59, but then I get errors in `deps_of_module` and `deps_of_vlib_module` because of the change, as some functions like `Modules.alias_for` still operate on `Module.t`.

Comment (jchavarri) [reply src/dune_rules/modules.mli]:
```
File "src/dune_rules/dep_rules.ml", line 62, characters 16-20:
62 |       aliases @ deps)
                     ^^^^
Error: This expression has type Modules.Sourced_module.t list
       but an expression was expected of type Module.t list
       Type Modules.Sourced_module.t is not compatible with type Module.t
```

Comment (rgrinberg) [reply src/dune_rules/modules.mli]: Right, but you can extract the `Module.t` out of the `Sourced_module.t` without too many issues.

Comment (jchavarri) [reply src/dune_rules/modules.mli]: This builds, but I'm not sure I got the idea right, as it seems more complex than before 🤔 https://github.com/ocaml/dune/pull/7187/commits/559e885f0af7f1d9b4bc1b590496e8a637195de2

Comment (rgrinberg) [reply src/dune_rules/modules.mli]: I think you can simplify it by keeping `ocamldep.mli` the same, and moving the `to_module` conversion to `ocamldep.ml`.

Comment (jchavarri) [reply src/dune_rules/modules.mli]: Tried something in https://github.com/ocaml/dune/pull/7187/commits/81d333991e480a7305a601fcfe6831146d06ebb1.

File: .github/workflows/bench.yml [hunk 9]
```diff
@@ -5,7 +5,7 @@ on:
   push:
     branches:
       - main
-      - bench/add-warm-run
+      - perf/run-parse_compilation_units-once
```
Comment (rgrinberg) [inline .github/workflows/bench.yml:8]: Should this be moved to a separate commit?

**Comment by rgrinberg** [comment]
I simplified things a little further. We don't need Obj_map at all.
